python/ray/data/grouped_data.py

```diff
@@ -297,19 +297,44 @@ def map_groups(
         else:
             sorted_ds = self._dataset.repartition(1)
 
-        # Returns the group boundaries.
-        def get_key_boundaries(block_accessor: BlockAccessor):
+        def get_key_boundaries(block_accessor: BlockAccessor) -> List[int]:
+            """Compute block boundaries based on the key(s)"""
+
             import numpy as np
 
-            boundaries = []
             # Get the keys of the batch in numpy array format
             keys = block_accessor.to_numpy(self._key)
-            start = 0
-            while start < keys.size:
-                end = start + np.searchsorted(keys[start:], keys[start], side="right")
-                boundaries.append(end)
-                start = end
-            return boundaries
+
+            has_single_key = isinstance(keys, np.ndarray)
+
+            if has_single_key:
+                boundaries = []
+                start = 0
+                while start < len(keys):
+                    end = start + np.searchsorted(
+                        keys[start:], keys[start], side="right"
+                    )
+                    boundaries.append(end)
+                    start = end
+                return boundaries
+
+            else:
+                # For multiple keys, we generate a separate tuple column
+                # Due to how np functions treats tuples, we need to cast
+                # it into an array of tuple
+                first_key = next(iter(keys))
+                arr = np.empty(len(keys[first_key]), dtype=object)
+                arr[:] = list(zip(*keys.values()))
+
+                boundaries = []
+                start = 0
+                while start < len(arr):
+                    x = np.empty(1, dtype=object)
+                    x[0] = arr[start]
```

Comment:
What's going on here? Also, could we rename `x` with a more descriptive name?

Reply:
After [constructing an array of tuples](https://github.com/ray-project/ray/pull/40778/files#diff-cbb5f0a6b5da217cf7b96fdba5bba7f534871e87a56fb3349380395a47a2ec99R327), the target value `x` of the `np.searchsorted` needs to be a single-element tuple ndarray. Otherwise, `searchsorted` complains about the missing `<` operator. Alternatively, we can add a small class with `__lt__` method instead of using tuples for this PR.